Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,207 @@
+        // TODO: Make developers give us absolute path
+        // or, take the base path as an argument
+        return { -1, Optional<IPC::File> {} };
+    }
+
+    StringBuilder builder;
+    if (requested_access & FileAccess::Read)
+        builder.append('r');
+
+    if (requested_access & FileAccess::Write)
+        builder.append('w');
+
+    if (requested_access & FileAccess::CreateOrRemove)
+        builder.append('c');
+
+    auto access_string = builder.to_string();
+
+    auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} \"{}\" access to \"{}\"?", sender, access_string, path), "File Permissions Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+    if (result == GUI::MessageBox::ExecYes) {
+        int access = 0;
+
+        if (requested_access & FileAccess::Read)
+            access |= O_RDONLY;
+
+        if (requested_access & FileAccess::Write)
+            access |= O_WRONLY;
+
+        if (requested_access & FileAccess::CreateOrRemove)
+            access |= O_CREAT;
+
+        if (requested_access & FileAccess::Truncate)
+            access |= O_TRUNC;
+
+        auto fd = open(path.characters(), access);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+
+            // TODO: check for root? or 1000
+
+            return { errno, Optional<IPC::File> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    if (result == GUI::MessageBox::ExecNo) {
+        // TODO: If a malicious program requests access over and over, do we eventually want to deny it?
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptOpenFileResponse ClientConnection::prompt_open_file(String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+
+    Optional<String> user_picked_file = GUI::FilePicker::get_open_filepath(main_window, "Select file", path_to_view);
+
+    if (user_picked_file.has_value()) {
+        int access = 0;
+
+        if (requested_access & FileAccess::Read)
+            access |= O_RDONLY;
+
+        if (requested_access & FileAccess::Write)
+            access |= O_WRONLY;
+
+        if (requested_access & FileAccess::CreateOrRemove)
+            access |= O_CREAT;
+
+        auto fd = open(user_picked_file.value().characters(), access, 0644);
+
+        if (fd < 0) {
+            dbgln(OSError(errno).string());
+
+            // TODO: check for root? or 1000
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+Messages::FileSystemAccessServer::PromptSaveFileResponse ClientConnection::prompt_save_file(String const& name, String const& ext, String const& path_to_view, int requested_access)
```

Comment:
There is a lot of code duplication between `prompt_open_file` and `prompt_save_file`, maybe you could add a helper method for everything after the `FilePicker` dialog invocation, since that code seems to be identical.

Reply:
This should be less of a problem now that I've taken out the custom access flag logic, I think it will make sense to keep these separate until they bulk up again.

Reply:
Done.